### PR TITLE
More NonEmpty functions

### PR DESCRIPTION
**Pull request by leonidr-da**
### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by leonidr-da** [comment]
I was also going to add a `fromList` to a `DA.NonEmpty.Total` but per https://discuss.daml.com/t/to-abort-or-to-fail/3012/8, I'm not certain whether we should ActionFail or CanAbort.

File: compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml [hunk 1]
```diff
@@ -99,6 +106,28 @@ toList (NonEmpty head tail) = head :: tail
...
 
+-- | Find an element in a non-empty list.
+find : (a -> Bool) -> NonEmpty a -> Optional a
+find p (NonEmpty head tail)
+  | p head = Some head
+  | otherwise = L.find p tail
+
+-- | The 'deleteBy' function behaves like 'delete', but takes a
+-- user-supplied equality predicate.
+deleteBy : (a -> a -> Bool) -> a -> NonEmpty a -> Optional (NonEmpty a)
```
Comment (cocreature) [inline compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: Have you considered returning `[]` instead? `Optional (NonEmpty a)` doesn’t provide any extra info or type safety.

Comment (leonidr-da) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: Good point. I had not considered it. I think given the  trade-off of unwrapping an `Optional` or recreating the `NonEmpty` I prefer the former. Handling the case where you delete the last element is rarer than deleting an element or deleting none. But, yes, that type does look a bit weird.

Comment (leonidr-da) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: @cocreature Is this ok with you? It would be nice to get this done.

Comment (cocreature) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: sorry missed that, I think I have a preference for the `[a]` case. The type signature here just looks confusing.

Comment (leonidr-da) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: Can you say why it looks confusing? After this `delete` one will have to handle either a `None` or a `[]`, and there are plenty of [examples](https://docs.daml.com/daml/stdlib/DA-List.html#function-da-list-elemindex-66976) where we wrap in `Optional` to deal with the "missing" case.

Comment (cocreature) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml:117]: Generally, I opt for the simplest type that expresses the properties I care about. In this case `[a]` is objectively simpler than `Optional (NonEmpty a)` (both at the type and at the value level) but the more complex type provides zero extra information.

While Haskell doesn’t have `delete` on `NonEmpty` it has [filter](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List-NonEmpty.html#v:filter) which is close enough and follows my suggestion here.

Comment (leonidr-da) [reply compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml]: I don't think that simplicity is the only thing one should consider. I do think that  `Optional (NonEmpty a)` is more explicit in  expressing to the user that some operation may not return a `NonEmpty a`.  But I won't belabor this.

**Review by cocreature — APPROVED** [review]
Nice, thank you!

**Pull request merged**